File: src/lib/puzzle.ts
export const UNKNOWN = -1;
export const BLACK   =  0;  // water
export const WHITE   =  1;  // island

export class Island {
  id:     number;
  origin: [number, number];
  size:   number;
  cells:  Set<string>;  // "r,c" string keys

  constructor(origin: [number, number], size: number, id: number) {
    this.id     = id;
    this.origin = origin;
    this.size   = size;
    this.cells  = new Set([`${origin[0]},${origin[1]}`]);
  }
}

export class Grid {
  rows:     number;
  cols:     number;
  clues:    Map<string, number>;  // "r,c" -> island size
  cells:    number[];             // UNKNOWN / BLACK / WHITE per cell
  islandId: number[];             // island id per cell, -1 = none
  islands:  Island[];

  constructor(rows: number, cols: number, clues: Map<string, number>) {
    this.rows     = rows;
    this.cols     = cols;
    this.clues    = clues;
    const n       = rows * cols;
    this.cells    = new Array(n).fill(UNKNOWN);
    this.islandId = new Array(n).fill(-1);
    this.islands  = [];
    this._buildIslands();
  }

  private _idx(r: number, c: number): number { return r * this.cols + c; }

  private _buildIslands(): void {
    let iid = 0;
    for (const [key, sz] of this.clues) {
      const [r, c] = key.split(',').map(Number);
      const isl    = new Island([r, c], sz, iid);
      this.islands.push(isl);
      const idx = this._idx(r, c);
      this.cells[idx]    = WHITE;
      this.islandId[idx] = iid;
      iid++;
    }
  }

  get(r: number, c: number): number          { return this.cells[this._idx(r, c)]; }
  set(r: number, c: number, v: number): void { this.cells[this._idx(r, c)] = v; }

  getIslandId(r: number, c: number): number          { return this.islandId[this._idx(r, c)]; }
  setIslandId(r: number, c: number, id: number): void { this.islandId[this._idx(r, c)] = id; }

  neighbors(r: number, c: number): [number, number][] {
    const res: [number, number][] = [];
    if (r > 0)              res.push([r - 1, c]);
    if (r < this.rows - 1) res.push([r + 1, c]);
    if (c > 0)              res.push([r, c - 1]);
    if (c < this.cols - 1) res.push([r, c + 1]);
    return res;
  }

  copy(): Grid {
    const g      = Object.create(Grid.prototype) as Grid;
    g.rows       = this.rows;
    g.cols       = this.cols;
    g.clues      = this.clues;                    // immutable, shared
    g.cells      = [...this.cells];
    g.islandId   = [...this.islandId];
    g.islands    = this.islands.map(isl => {
      const c    = new Island(isl.origin, isl.size, isl.id);
      c.cells    = new Set(isl.cells);
      return c;
    });
    return g;
  }

  restore(snap: Grid): void {
    this.cells    = [...snap.cells];
    this.islandId = [...snap.islandId];
    for (let i = 0; i < snap.islands.length; i++)
      this.islands[i].cells = new Set(snap.islands[i].cells);
  }

  isComplete(): boolean { return this.cells.every(v => v !== UNKNOWN); }
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

export function parsePuzzle(text: string): Grid {
  const rowsData: number[][] = [];
  for (let line of text.split('\n')) {
    line = line.trim();
    if (!line || line.startsWith('#')) continue;
    line = line.replace(/\./g, '0');
    const tokens = line.split(/[\s,]+/).filter(Boolean);
    if (!tokens.length) continue;
    const row: number[] = [];
    for (const t of tokens) {
      const n = parseInt(t, 10);
      if (isNaN(n)) throw new Error(`Invalid token '${t}'`);
      row.push(n);
    }
    if (row.length) rowsData.push(row);
  }

  if (!rowsData.length) throw new Error('Empty puzzle');
  const numCols = rowsData[0].length;
  for (let i = 0; i < rowsData.length; i++)
    if (rowsData[i].length !== numCols)
      throw new Error(`Row ${i} has ${rowsData[i].length} cols, expected ${numCols}`);

  const clues = new Map<string, number>();
  for (let r = 0; r < rowsData.length; r++)
    for (let c = 0; c < rowsData[r].length; c++) {
      const v = rowsData[r][c];
      if (v < 0) throw new Error(`Negative clue at (${r},${c})`);
      if (v > 0) clues.set(`${r},${c}`, v);
    }

  if (!clues.size) throw new Error('No island clues found');
  return new Grid(rowsData.length, numCols, clues);
}

// ---------------------------------------------------------------------------
// Validator
// ---------------------------------------------------------------------------

export function validateSolution(grid: Grid): [boolean, string] {
  const { rows, cols } = grid;

  // 1. No UNKNOWN cells
  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++)
      if (grid.get(r, c) === UNKNOWN) return [false, 'Grid has unknown cells'];

  // 2. Each island clue correct size, no overlap
  const visitedWhite = new Set<string>();
  for (const [key, sz] of grid.clues) {
    const [cr, cc] = key.split(',').map(Number);
    if (grid.get(cr, cc) !== WHITE) return [false, `Clue (${cr},${cc}) is not white`];
    const q: [number, number][] = [[cr, cc]];
    let qi = 0;
    const isle = new Set<string>();
    while (qi < q.length) {
      const [r, c] = q[qi++];
      const k = `${r},${c}`;
      if (isle.has(k)) continue;
      isle.add(k);
      for (const [nr, nc] of grid.neighbors(r, c))
        if (grid.get(nr, nc) === WHITE && !isle.has(`${nr},${nc}`))
          q.push([nr, nc]);
    }
    if (isle.size !== sz)
      return [false, `Island at (${cr},${cc}) has ${isle.size} cells, expected ${sz}`];
    for (const cell of isle)
      if (visitedWhite.has(cell)) return [false, `Islands merged at ${cell}`];
    for (const cell of isle) visitedWhite.add(cell);
  }

  // 3. All white cells belong to an island
  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++)
      if (grid.get(r, c) === WHITE && !visitedWhite.has(`${r},${c}`))
        return [false, `White cell (${r},${c}) not part of any island`];

  // 4. No 2×2 black block
  for (let r = 0; r < rows - 1; r++)
    for (let c = 0; c < cols - 1; c++)
      if ([[0,0],[0,1],[1,0],[1,1]].every(([dr,dc]) => grid.get(r+dr, c+dc) === BLACK))
        return [false, `2×2 black block at (${r},${c})`];

  // 5. Black cells connected
  const blacks: [number, number][] = [];
  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++)
      if (grid.get(r, c) === BLACK) blacks.push([r, c]);
  if (!blacks.length) return [false, 'No black cells'];

  const vb = new Set<string>();
  const bq: [number, number][] = [blacks[0]];
  let bqi = 0;
  while (bqi < bq.length) {
    const [r, c] = bq[bqi++];
    const k = `${r},${c}`;
    if (vb.has(k)) continue;
    vb.add(k);
    for (const [nr, nc] of grid.neighbors(r, c))
      if (grid.get(nr, nc) === BLACK && !vb.has(`${nr},${nc}`))
        bq.push([nr, nc]);
  }
  if (vb.size !== blacks.length) return [false, 'Black cells are not all connected'];

  return [true, 'Valid solution'];
}

File: puzzle.py
"""Grid model, file parser, and solution validator for Nurikabe."""

from __future__ import annotations
from collections import deque

UNKNOWN = -1
BLACK   =  0   # water
WHITE   =  1   # island


class Island:
    def __init__(self, origin: tuple[int, int], size: int, island_id: int):
        self.origin    = origin
        self.size      = size
        self.id        = island_id
        self.cells: set[tuple[int, int]] = {origin}

    def __repr__(self) -> str:
        return f"Island(id={self.id}, origin={self.origin}, size={self.size}, cells={len(self.cells)})"


class Grid:
    def __init__(self, rows: int, cols: int, clues: dict[tuple[int, int], int]):
        self.rows   = rows
        self.cols   = cols
        self.clues  = clues  # {(r,c): size}

        n = rows * cols
        # -1 = UNKNOWN, 0 = BLACK, 1 = WHITE
        self.cells: bytearray = bytearray(b'\xff' * n)  # 0xff == 255 used as sentinel; we map below
        # Use signed-style: store as 0=BLACK,1=WHITE,2=UNKNOWN internally in bytearray
        # Convention: 0->BLACK, 1->WHITE, 2->UNKNOWN
        self.cells = bytearray([2] * n)  # 2 == UNKNOWN sentinel

        # island id per cell; -1 = none
        self.island_id: list[int] = [-1] * n

        # Islands indexed by id
        self.islands: list[Island] = []
        self._build_islands()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------
    def _idx(self, r: int, c: int) -> int:
        return r * self.cols + c

    def _build_islands(self) -> None:
        for iid, ((r, c), sz) in enumerate(self.clues.items()):
            isl = Island((r, c), sz, iid)
            self.islands.append(isl)
            idx = self._idx(r, c)
            self.cells[idx] = 1  # WHITE
            self.island_id[idx] = iid

    # ------------------------------------------------------------------
    # Public cell accessors
    # ------------------------------------------------------------------
    def get(self, r: int, c: int) -> int:
        """Return UNKNOWN(-1), BLACK(0), or WHITE(1)."""
        v = self.cells[self._idx(r, c)]
        return -1 if v == 2 else v  # 2->UNKNOWN(-1), 0->BLACK, 1->WHITE

    def set(self, r: int, c: int, value: int) -> None:
        """Set cell to UNKNOWN(-1), BLACK(0), or WHITE(1)."""
        self.cells[self._idx(r, c)] = 2 if value == UNKNOWN else value

    def get_island_id(self, r: int, c: int) -> int:
        return self.island_id[self._idx(r, c)]

    def set_island_id(self, r: int, c: int, iid: int) -> None:
        self.island_id[self._idx(r, c)] = iid

    def neighbors(self, r: int, c: int) -> list[tuple[int, int]]:
        result = []
        if r > 0:             result.append((r - 1, c))
        if r < self.rows - 1: result.append((r + 1, c))
        if c > 0:             result.append((r, c - 1))
        if c < self.cols - 1: result.append((r, c + 1))
        return result

    # ------------------------------------------------------------------
    # Snapshot / copy
    # ------------------------------------------------------------------
    def copy(self) -> Grid:
        g = Grid.__new__(Grid)
        g.rows      = self.rows
        g.cols      = self.cols
        g.clues     = self.clues  # immutable, shared
        g.cells     = bytearray(self.cells)
        g.island_id = list(self.island_id)
        g.islands   = [Island(isl.origin, isl.size, isl.id) for isl in self.islands]
        for isl_orig, isl_copy in zip(self.islands, g.islands):
            isl_copy.cells = set(isl_orig.cells)
        return g

    def restore(self, snapshot: Grid) -> None:
        self.cells[:]     = snapshot.cells
        self.island_id[:] = snapshot.island_id
        for i, isl in enumerate(snapshot.islands):
            self.islands[i].cells = set(isl.cells)

    # ------------------------------------------------------------------
    # Status
    # ------------------------------------------------------------------
    def is_complete(self) -> bool:
        for r in range(self.rows):
            for c in range(self.cols):
                if self.get(r, c) == UNKNOWN:
                    return False
        return True

    def count_unknown(self) -> int:
        return sum(1 for v in self.cells if v == 2)

    # ------------------------------------------------------------------
    # Debug
    # ------------------------------------------------------------------
    def __str__(self) -> str:
        chars = {UNKNOWN: '.', BLACK: '#', WHITE: 'O'}
        rows_str = []
        for r in range(self.rows):
            row = []
            for c in range(self.cols):
                v = self.get(r, c)
                if v == WHITE and (r, c) in self.clues:
                    row.append(str(self.clues[(r, c)]))
                else:
                    row.append(chars[v])
            rows_str.append(' '.join(row))
        return '\n'.join(rows_str)


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------

def parse_puzzle(text: str) -> Grid:
    """Parse puzzle text into a Grid. Raises ValueError on bad input."""
    rows_data = []
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        # Replace dots with 0
        line = line.replace('.', '0')
        # Split on whitespace or commas
        import re
        tokens = re.split(r'[\s,]+', line)
        if not tokens:
            continue
        row = []
        for t in tokens:
            if t == '':
                continue
            try:
                row.append(int(t))
            except ValueError:
                raise ValueError(f"Invalid token '{t}' in puzzle")
        if row:
            rows_data.append(row)

    if not rows_data:
        raise ValueError("Empty puzzle")

    num_rows = len(rows_data)
    num_cols = len(rows_data[0])
    for i, row in enumerate(rows_data):
        if len(row) != num_cols:
            raise ValueError(f"Row {i} has {len(row)} cols, expected {num_cols}")

    clues: dict[tuple[int, int], int] = {}
    for r, row in enumerate(rows_data):
        for c, val in enumerate(row):
            if val < 0:
                raise ValueError(f"Negative clue at ({r},{c})")
            if val > 0:
                clues[(r, c)] = val

    if not clues:
        raise ValueError("No island clues found")

    return Grid(num_rows, num_cols, clues)


# ---------------------------------------------------------------------------
# Validator
# ---------------------------------------------------------------------------

def validate_solution(grid: Grid) -> tuple[bool, str]:
    
    """
    Check if the grid is a valid Nurikabe solution.
    Returns (True, "Valid solution") or (False, reason).
    """
    rows, cols = grid.rows, grid.cols

    # 1. No UNKNOWN cells
    for r in range(rows):
        for c in range(cols):
            if grid.get(r, c) == UNKNOWN:
                return False, "Grid has unknown cells"

    # 2. Each island clue has correct total size
    #    Use BFS/DFS from clue cells
    visited_white: set[tuple[int, int]] = set()

    for (cr, cc), sz in grid.clues.items():
        if grid.get(cr, cc) != WHITE:
            return False, f"Clue cell ({cr},{cc}) is not WHITE"
        # BFS flood fill this island
        q = deque([(cr, cc)])
        island_cells = set()
        while q:
            r, c = q.popleft()
            if (r, c) in island_cells:
                continue
            island_cells.add((r, c))
            for nr, nc in grid.neighbors(r, c):
                if grid.get(nr, nc) == WHITE and (nr, nc) not in island_cells:
                    q.append((nr, nc))
        # Island must be exactly sz cells
        if len(island_cells) != sz:
            return False, f"Island at ({cr},{cc}) has {len(island_cells)} cells, expected {sz}"
        # Must not overlap another clue's island (they'd merge)
        for cell in island_cells:
            if cell in visited_white:
                return False, f"Islands are connected (merged at {cell})"
        visited_white |= island_cells

    # 3. All WHITE cells belong to some island
    for r in range(rows):
        for c in range(cols):
            if grid.get(r, c) == WHITE and (r, c) not in visited_white:
                return False, f"WHITE cell ({r},{c}) not part of any clue island"

    # 4. No 2×2 block of BLACK
    for r in range(rows - 1):
        for c in range(cols - 1):
            if all(grid.get(r + dr, c + dc) == BLACK
                   for dr in (0, 1) for dc in (0, 1)):
                return False, f"2×2 BLACK block at ({r},{c})"

    # 5. BLACK cells form a single connected component
    black_cells = [(r, c) for r in range(rows) for c in range(cols)
                   if grid.get(r, c) == BLACK]
    if not black_cells:
        return False, "No BLACK cells"
    visited_black: set[tuple[int, int]] = set()
    q = deque([black_cells[0]])
    while q:
        r, c = q.popleft()
        if (r, c) in visited_black:
            continue
        visited_black.add((r, c))
        for nr, nc in grid.neighbors(r, c):
            if grid.get(nr, nc) == BLACK and (nr, nc) not in visited_black:
                q.append((nr, nc))
    if len(visited_black) != len(black_cells):
        return False, "BLACK (water) cells are not all connected"

    return True, "Valid solution"
